cylc/flow/scheduler.py

```diff
@@ -1431,12 +1405,66 @@ async def main_loop(self):
                     self.data_store_mgr.publish_deltas)
 
             self.process_command_queue()
+
             if not self.is_paused:
-                self.release_runahead_tasks()
+                if self.pool.release_runahead_tasks():
+                    self.is_updated = True
+                    self.reset_inactivity_timer()
+
             self.proc_pool.process()
 
-            if self.should_process_tasks():
-                self.process_task_pool()
+            # Get tasks with newly-satisfied external triggers.
+            check_if_ready = self.broadcast_mgr.check_ext_triggers(
+                [x for x in self.pool.get_tasks()
```

Comment:
Yeah, we have discussed reformulating the task pool to a dict (it's mentioned in an issue somewhere) to avoid some iteration.  Beyond the scope of this PR though!

Reply:
(Reformulating the task pool is definitely on the table, as a possible future enhancement).